control-plane/connect-inject/terminatinggatewayservice_controller.go

```diff
@@ -0,0 +1,440 @@
+
+	policyName := ""
+	if r.AclEnabled {
+		policyName = terminatingGatewayService.Spec.Service.ServiceName + "-write-policy"
+	}
+
+	terminatingGatewayService.Status.LastSyncedTime = &metav1.Time{Time: time.Now()}
+	terminatingGatewayService.SetSyncedCondition(corev1.ConditionTrue, "", "")
+
+	terminatingGatewayService.Status.ServiceInfoRef = &consulv1alpha1.ServiceInfoRefStatus{
+		ServiceName: terminatingGatewayService.Spec.Service.ServiceName,
+		PolicyName:  policyName,
+	}
+
+	err := r.Status().Update(ctx, terminatingGatewayService)
+	if err != nil {
+		r.Log.Error(err, "failed to update TerminatingGatewayService status", "name", terminatingGatewayService.Name, terminatingGatewayService.Namespace)
+	}
+	return err
+}
+
+// updateStatusError updates the terminatingGatewayService's ReconcileError in the status.
+func (r *TerminatingGatewayServiceController) updateStatusError(ctx context.Context, terminatingGatewayService *consulv1alpha1.TerminatingGatewayService, reconcileErr error) {
+	terminatingGatewayService.SetSyncedCondition(corev1.ConditionFalse, "Error updating status", reconcileErr.Error())
+
+	err := r.Status().Update(ctx, terminatingGatewayService)
+	if err != nil {
+		r.Log.Error(err, "failed to update TerminatingGatewayService status", "name", terminatingGatewayService.Name, terminatingGatewayService.Namespace)
+	}
+
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *TerminatingGatewayServiceController) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&consulv1alpha1.TerminatingGatewayService{}).
+		Complete(r)
+}
+
+func (r *TerminatingGatewayServiceController) createOrUpdateService(terminatingGatewayService *consulv1alpha1.TerminatingGatewayService, ctx context.Context) error {
+	spec := terminatingGatewayService.Spec
+
+	service, serviceExists, err := r.serviceFound(spec.Service.ServiceName)
+	if err != nil {
+		r.Log.Error(err, "Error obtaining existing services")
+		return err
+	}
+
+	if !serviceExists {
+		// register external service with Consul.
+		catalogRegistration := &api.CatalogRegistration{
+			Node:    spec.Service.Node,
+			Address: spec.Service.Address,
+			Service: &api.AgentService{
+				ID:      spec.Service.ServiceName,
+				Service: spec.Service.ServiceName,
+				Port:    spec.Service.ServicePort}}
+
+		_, err = r.ConsulClient.Catalog().Register(catalogRegistration, nil)
+		if err != nil {
+			r.Log.Error(err, "Unable to register external service with Consul")
+			return err
+		}
+
+		if r.AclEnabled {
+			err := r.updateTerminatingGatewayTokenWithWritePolicy(terminatingGatewayService)
+			if err != nil {
+				r.Log.Error(err, "Unable to update the terminating gateway ACL token with new write policy")
+				return err
+			}
+		}
+
+		// Store the state in the status.
+		err = r.updateStatus(ctx, terminatingGatewayService)
+		return err
+	}
+
+	err = r.updateServiceIfDifferent(service, terminatingGatewayService, ctx)
```

Comment:
Would we need to update the ACL token at all if we are calling updateServiceIfDifferent? Also should this be in an "else" clause or do we always want to run this even if we just newly created the service?

Reply:
There is an implicit else in the code. For a newly created service, we would return from the function on line 191 before `updateServiceIfDifferent`

Reply:
Aha I see, that makes sense.